server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -999,6 +986,26 @@ void createRetentionLease(final long startingSeqNo, ActionListener<RetentionLeas
         }), shard, cancellableThreads, leaseListener.delegateResponse((l, e) -> outerListener.onFailure(e)));
     }
 
+    private void deleteRetentionLease(ActionListener<Void> outerListener) {
+        // NB we release the operation permit as soon as we have deleted the lease, but delay the outer listener until it is synced
+        final var leaseListener = new SubscribableListener<Void>();
+        final var delayedListener = new ThreadedActionListener<>(
+            shard.getThreadPool().generic(),
+            outerListener.<Void>delegateFailure((l, ignored) -> leaseListener.addListener(l))
+        );
+
+        runUnderPrimaryPermit(permitListener -> ActionListener.completeWith(permitListener, () -> {
+            try {
+                shard.removePeerRecoveryRetentionLease(request.targetNode().getId(), delayedListener.map(ignored -> null));
+            } catch (RetentionLeaseNotFoundException e) {
+                // this counts as success
+                logger.debug("no peer-recovery retention lease for " + request.targetAllocationId());
```

Comment:
NIT :) Not that it really matters and it was like that before but maybe remove string concat from logger invocation here?

Reply:
Ah yep ok, see 87ca673238a